### PR TITLE
feat: changes for user management

### PR DESCRIPTION
**Pull request by prashbnair**

### CODE REVIEW

File: bayesian/utility/user_utils.py
Comment (deepak1725) [inline bayesian/utility/user_utils.py:34]: Can we move this to `db_gateway.py`

File: bayesian/api/user_api.py
Comment (deepak1725) [inline bayesian/api/user_api.py:41]: I think `user_id` will throw `KeyError` if user_id is not in request. Can you please check.

Comment (deepak1725) [inline bayesian/api/user_api.py:41]:
```suggestion
    user_id = content.get('user_id')
```

File: bayesian/api/user_api.py
Comment (deepak1725) [inline bayesian/api/user_api.py:46]: This may again Throw `KeyError` if `snyk_api_token` not in request.

```suggestion
    snyk_api_token = content.get('snyk_api_token')
```

File: bayesian/api/user_api.py
Comment (deepak1725) [inline bayesian/api/user_api.py:55]: I suggest to do it in this way. Looks cleaner
```suggestion
if not user_utils.is_snyk_token_valid(snyk_api_token):
     raise HTTPError(400, "Invalid API Token")
 encrypted_api_token = user_utils.encrypt_api_token(snyk_api_token)
 user_utils.create_or_update_user(user_id, encrypted_api_token.decode(), "SNYK")
```

File: openshift/template.yaml
Comment (deepak1725) [inline openshift/template.yaml:104]: Short, Precise, Easy to read
```suggestion
                key: snyk_encryption_token
```

File: bayesian/utility/user_utils.py
Comment (arajkumar) [inline bayesian/utility/user_utils.py:75]: instead of using single key to encrypt all Snyk API Keys, I suggest to derive a key from UUID for each API key insertion. Something like below,

```
store(uuid, snyk_api_key):
  uuid_hash = sha256(uuid)
  snyk_api_key_encrypted = AES(data=snyk_api_key, key=uuid + salt)
  snyk_api_key_hash = sha256(snyk_api_key)

read(uuid):
  snyk_api_key = AES_decrypt(data=snyk_api_key_encrypted, key=uuid + salt)
```

This approach introduces the following columns to the user database, 
```
uuid_hash - SHA256 hashed value of original UUID
synk_api_key_hash - SHA256 hashed value of original Synk API key to check existence of key without decryption
snyk_api_key_encrypted - AES encrypted Synk API key based original UUID and salt which can be used to query snyk API
```
Also the encryption key for each Snyk API token is different and based on UUID., this ensures that compromise of single encryption key will not expose the rest of the user data as well.

File: bayesian/utility/user_utils.py
Comment (arajkumar) [inline bayesian/utility/user_utils.py:37]: Isn't it redundant validation? Request has been already validated by the route handler right?

File: bayesian/utility/user_utils.py
Comment (arajkumar) [inline bayesian/utility/user_utils.py:66]:
```suggestion
        return response.status_code == 200
```

File: bayesian/utility/user_utils.py
Comment (arajkumar) [inline bayesian/utility/user_utils.py:62]: `Content-Type` is automatically implied to `application/json` if json named argument is given to requests.post.

File: bayesian/utility/user_utils.py
Comment (arajkumar) [inline bayesian/utility/user_utils.py:16]:
```suggestion
from bayesian.default_config import SNYK_API_TOKEN_VALIDATION_URL, ENCRYPTION_KEY_FOR_SNYK_TOKEN
```

File: bayesian/utility/user_utils.py
Comment (arajkumar) [inline bayesian/utility/user_utils.py:63]: Usual pattern is to use `response.raise_for_status()`

File: bayesian/utility/user_utils.py
Comment (arajkumar) [inline bayesian/utility/user_utils.py:69]: How about using Exception for flow control? it is quite common in Python world.
The given token is valid if there no Exeption, otherwise invalid.

```suggestion
response = requests.post(SNYK_API_TOKEN_VALIDATION_URL, json={'api': snyk_api_token})
response.raise_for_status()
```

File: bayesian/api/user_api.py
Comment (prashbnair) [reply bayesian/api/user_api.py:41]: Done

File: bayesian/api/user_api.py
Comment (prashbnair) [reply bayesian/api/user_api.py:46]: Done

File: bayesian/api/user_api.py
Comment (prashbnair) [reply bayesian/api/user_api.py:55]: Done

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:16]: Done

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:34]: I am not sure if we want all DB interactions to go through that. That file would then being to become very big.

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:37]: Done

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:66]: Done

File: bayesian/utility/user_utils.py
Comment (deepak1725) [reply bayesian/utility/user_utils.py:34]: I leave it up to you. :)

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:75]: Can you let me know why we need to store the hash values. Hashing is one way and  the value cannot be retrieved. There is also no comparison to be done against this value so why should it be stored.
If you are using a salt dont you need to remember the salt for decryption.

File: openshift/template.yaml
Comment (prashbnair) [reply openshift/template.yaml:104]: It is an encryption key for the API token.

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:62]: Done

File: bayesian/utility/user_utils.py
Comment (arajkumar) [reply bayesian/utility/user_utils.py:75]: > Can you let me know why we need to store the hash values

Hashing the Synk API key is meant for comparison, if you want to check whether the Snyk API key is already exists during a call to create_or_update user. Also this can be used to report how many users are using same Snyk API key.

> If you are using a salt dont you need to remember the salt for decryption

Yes, salt should be a global configuration(as an env variable). So compromising either salt or uuid can't expose Snyk API key from DB dump.

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:75]: I am not convinced that this a use case for hashing. We already store the encrypted form and there is no need to do both. For reporting we can always decrypt and figure out how may users are using the same key.

My feeling is that once the salt is leaked it is only a matter of time before it is decrypted. Salts are generally used to ensure that there is no collision when using hashing.

File: openshift/template.yaml
Comment (deepak1725) [reply openshift/template.yaml:104]: ok, Can we summarize it in 3 letter variable name..?

File: bayesian/utility/user_utils.py
Comment (arajkumar) [reply bayesian/utility/user_utils.py:75]: Leave about Hashing, what is the problem in using UUID as an encryption key instead of global encryption key?

Comment (prashbnair) [reply bayesian/utility/user_utils.py:75]: It just seems like the most obvious key that everybody will use if they want to decrypt.

File: bayesian/api/user_api.py
Comment (prashbnair) [reply bayesian/api/user_api.py:41]: Done

File: bayesian/utility/user_utils.py
Comment (arajkumar) [reply bayesian/utility/user_utils.py:75]: Though UUID will be used as key it won't be stored as a plain text in DB., UUID should be hashed and stored.

```
store(uuid, snyk_api_key):
  uuid_hash = sha256(uuid)
  snyk_api_key_encrypted = AES(data=snyk_api_key, key=uuid)

read(uuid):
  snyk_api_key = AES_decrypt(data=snyk_api_key_encrypted, key=uuid)
```

So even if the DB is compromised, data can't be decrypted unless you know each and every UUID.

Comment (prashbnair) [reply bayesian/utility/user_utils.py:75]: Whats the problem with storing the key as a secret. Thats exactly what it is for. If the secrets get compromised, then we nothing we add there is safe.

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:69]: I am not sure I understand this, response.raise_for_status is fine for cases where the request is made. What about the case when the request failed, I would like to print out a log saying the request has failed.

File: bayesian/utility/user_utils.py
Comment (arajkumar) [reply bayesian/utility/user_utils.py:75]: I'm not talking about the key compromising scenario. If database compromised, having a single key for all data would expose the complete data. However when using UUID, each API key will be encrypted using unique key and it broadly reduces the attack surface.

File: openshift/template.yaml
Comment (prashbnair) [reply openshift/template.yaml:104]: Is this a standard that we need to follow?

File: bayesian/utility/user_utils.py
Comment (prashbnair) [reply bayesian/utility/user_utils.py:69]: If the key cannot be compromised how can it be decrypted?  I am not seeing any additional value that this can provide.

File: openshift/template.yaml
Comment (deepak1725) [reply openshift/template.yaml:104]: 3 letter looks more readable..! If we promote readiabiility, then it can be good standard to follow..

File: bayesian/utility/user_utils.py
Comment (arajkumar) [reply bayesian/utility/user_utils.py:69]: Encrypted data can be compromised using brute force attack. 
With the existing approach, once the attacker gains the key through brute force, then the entire table will be compromised. However in the proposed solution, brute force should be applied for individual rows and would take time.

Comment (prashbnair) [reply bayesian/utility/user_utils.py:69]: I understand all of this, for this solution I am not convinced that we need to go through all of this. What we have is good enough for the problem that we have.

File: openshift/template.yaml
Comment (prashbnair) [reply openshift/template.yaml:104]: Isn't readability about clearly expressing what the variable means?  The only name that I can think of is encryption_key_snyk_token.

File: bayesian/default_config.py
Comment (yzainee) [inline bayesian/default_config.py:51]: I think you can add this to the snyk secrets

File: bayesian/utility/user_utils.py
Comment (arajkumar) [reply bayesian/utility/user_utils.py:69]: There is a value in data and that is the prime reason for encrypting it. When we choose to encrypt it, why not a hardened approach?. I'm not convinced in going ahead with the existing approach. Thanks.

Comment (yzainee) [reply bayesian/utility/user_utils.py:69]: @prashbnair i dont think we need so much of a sophistication for storing the snyk api key. A simple encryption mechanism should be good enough

**Review by yzainee — CHANGES_REQUESTED** [review]

File: openshift/template.yaml
Comment (deepak1725) [reply openshift/template.yaml:104]: Fine. Settling with `encryption_key_for_snyk_token` then

File: bayesian/api/user_api.py
Comment (prashbnair) [reply bayesian/api/user_api.py:29]: It is basically what the meaning that POST conveys in REST  which is create. I know we are not creating a user but we are generating a ID. GET ideally should not be doing any of this.

File: bayesian/default_config.py
Comment (prashbnair) [reply bayesian/default_config.py:51]: Ok

**Review by deepak1725 — CHANGES_REQUESTED** [review]
@prashbnair I have 2 Open Points only if you can let know your take on this, it will be helpful

1. `user_utils.py` don't match our v2 Standards, Why don't we considering shifting v2 code in new standardized named file..?

V2 as a folder exists because we have v2 api's . For the user api, as of now it is not versioned so it stands on its own.

2. I see a Query in `user_utils.py` in `create_or_update_user`.. Why we are not willing to shift it `db_gateway.py`

User Management is a small module with not too much functionality. In case of the API's there is a lot more functionality and it made sense to have a separate data access layer. If I need to move to this to db_gateway it will just be another indirection which is really not required if you look at the code complexity.

**Review by deepak1725 — APPROVED** [review]

File: openshift/template.yaml
Comment (arajkumar) [inline openshift/template.yaml:187]: This should be reverted.

File: openshift/template.yaml
Comment (arajkumar) [inline openshift/template.yaml:114]: This should be reverted.

File: bayesian/utility/user_utils.py
Comment (arajkumar) [reply bayesian/utility/user_utils.py:69]: @yzainee Whatever I'm proposing is as simple as the one which is used by @prashbnair . The only difference is selection of single key vs value specific key(UUID).

**Review by arajkumar — APPROVED** [review]
I'm fine with the changes except the encryption part.

File: openshift/template.yaml
Comment (prashbnair) [reply openshift/template.yaml:187]: My mistake, will change

**Review by yzainee — APPROVED** [review]
You will need to update the vault secrets before you merge this PR

**Pull request merged**
